lib/rex/socket/ssl_tcp.rb

```diff
@@ -109,6 +109,7 @@ def initsock_with_ssl_version(params, version)
     end
 
     self.sslctx.options = OpenSSL::SSL::OP_ALL
+    self.sslctx.options |= OpenSSL::SSL::OP_LEGACY_SERVER_CONNECT
```

Comment:
We'll actually need something similar to fix the raw Net:HTTP calls in framework:

```ruby
OpenSSL::SSL::SSLContext::DEFAULT_PARAMS[:options] |= OpenSSL::SSL::OP_LEGACY_SERVER_CONNECT
```

Reply:
Which looks like folk are using Net HTTP for these scenarios https://github.com/rapid7/metasploit-framework/issues/8705

Reply:
@jmartin-r7 Any preference on this one? I'm not a fan of globally changing OpenSSL::SSL::SSLContext, as this could impact other things that I'm not aware of. Even if it was the previous behavior. 

It's required for Net HTTP to work as intended:

```ruby
require 'socket'
require 'net/http'
require 'uri'

# Flag which was removed from OP_ALL in OpenSSL 3
# OpenSSL::SSL::SSLContext::DEFAULT_PARAMS[:options] |= OpenSSL::SSL::OP_LEGACY_SERVER_CONNECT
# Alternatively - enabling the flag which appears in the error
# OpenSSL::SSL::SSLContext::DEFAULT_PARAMS[:options] |= OpenSSL::SSL::OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION

puts Net::HTTP.get(URI('https://www.courts.ie/'))
```

If we introduce it, I could add it to msfenv https://github.com/rapid7/metasploit-framework/blob/b06b39dc71181aa425ccfd07f7b29674b7090350/lib/msfenv.rb#L4-L8 - but that could potentially impact edgecase scenarios in Pro that I'm not aware of

Reply:
From docs [here](https://linux.die.net/man/3/ssl_set_options):
>Patched OpenSSL client and unpatched server.
>
>If the option SSL_OP_LEGACY_SERVER_CONNECT or SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION is set then initial connections and renegotiation between patched OpenSSL clients and unpatched servers succeeds. If neither option is set then initial connections to unpatched servers will fail.
The option SSL_OP_LEGACY_SERVER_CONNECT is currently set by default even though it has security implications: otherwise it would be impossible to connect to unpatched servers (i.e. all of them initially) and this is clearly not acceptable. Renegotiation is permitted because this does not add any additional security issues: during an attack clients do not see any renegotiations anyway.
>
>As more servers become patched the option SSL_OP_LEGACY_SERVER_CONNECT will not be set by default in a future version of OpenSSL.
>
>OpenSSL client applications wishing to ensure they can connect to unpatched servers should always set SSL_OP_LEGACY_SERVER_CONNECT
>
>OpenSSL client applications that want to ensure they can not connect to unpatched servers (and thus avoid any security issues) should always clear SSL_OP_LEGACY_SERVER_CONNECT using SSL_CTX_clear_options() or SSL_clear_options().
>
>The difference between the SSL_OP_LEGACY_SERVER_CONNECT and SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION options is that SSL_OP_LEGACY_SERVER_CONNECT enables initial connections and secure renegotiation between OpenSSL clients and unpatched servers only, while SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION allows initial connections and renegotiation between OpenSSL and unpatched clients or servers.

`SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION` might be the better option if we set globally to ensure web servers put up by modules or services are compatible with clients from as many targets as possible however maybe that also should be in the server config vs global.

Reply:
I'm still on the fence about flipping the `SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION` flag on; I can look at that separately - was just hoping to meet feature parity between openssl1 and 3 for now as a first pass :+1: